src/main/java/org/broadinstitute/consent/http/service/ElasticSearchService.java

```diff
@@ -0,0 +1,180 @@
+package org.broadinstitute.consent.http.service;
+
+import co.elastic.clients.elasticsearch.ElasticsearchClient;
+import co.elastic.clients.elasticsearch.core.BulkRequest;
+import co.elastic.clients.elasticsearch.core.BulkResponse;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import org.broadinstitute.consent.http.db.DacDAO;
+import org.broadinstitute.consent.http.db.DataAccessRequestDAO;
+import org.broadinstitute.consent.http.db.DatasetDAO;
+import org.broadinstitute.consent.http.db.UserDAO;
+import org.broadinstitute.consent.http.models.Dac;
+import org.broadinstitute.consent.http.models.Dataset;
+import org.broadinstitute.consent.http.models.DatasetProperty;
+import org.broadinstitute.consent.http.models.Study;
+import org.broadinstitute.consent.http.models.StudyProperty;
+import org.broadinstitute.consent.http.models.User;
+import org.broadinstitute.consent.http.models.UserRole;
+import org.broadinstitute.consent.http.models.elastic_search.DatasetTerm;
+import org.broadinstitute.consent.http.models.elastic_search.UserTerm;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ElasticSearchService {
+
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+  private final ElasticsearchClient esClient;
+  private final DatasetDAO datasetDAO;
+  private final DataAccessRequestDAO dataAccessRequestDAO;
+  private final DacDAO dacDAO;
+  private final UserDAO userDAO;
+  private final UseRestrictionConverter useRestrictionConverter;
+
+  public ElasticSearchService(
+      ElasticsearchClient esClient,
+      UseRestrictionConverter useRestrictionConverter,
+      DatasetDAO datasetDAO,
+      DataAccessRequestDAO dataAccessRequestDAO,
+      DacDAO dacDAO,
+      UserDAO userDAO) {
+    this.esClient = esClient;
+    this.useRestrictionConverter = useRestrictionConverter;
+    this.datasetDAO = datasetDAO;
+    this.dataAccessRequestDAO = dataAccessRequestDAO;
+    this.dacDAO = dacDAO;
+    this.userDAO = userDAO;
+  }
+
+  public BulkResponse indexDatasets(List<DatasetTerm> datasets) throws IOException {
+    BulkRequest.Builder br = new BulkRequest.Builder();
+
+    for (DatasetTerm dataset : datasets) {
+      br.operations(op -> op
+          .index(idx -> idx
+              .index("datasets")
```

Comment:
By check that the index exists, do you mean ensure that there is not a dataset with the ID provided already in elastic search? If so, [I think `index` is by default an `upsert` operation](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html): 

> `create` fails if a document with the same ID already exists in the target, `index` adds or replaces a document as necessary

Or, do you mean check that `datasets` is a valid elastic search index? I am not sure how to... The modern client's documentation never shows how to query the existence of an index or to create one. I think it handles that automatically. From what I can tell, I think it doesn't matter, because[ it will create the index if it doesn't exist already automatically](https://www.elastic.co/guide/en/elasticsearch/reference/8.7/docs-index_.html): 

> If the target doesn’t exist and matches the name or wildcard (*) pattern of an [index template with a data_stream definition](https://www.elastic.co/guide/en/elasticsearch/reference/8.7/set-up-a-data-stream.html#create-index-template), this request creates the data stream. See [Set up a data stream](https://www.elastic.co/guide/en/elasticsearch/reference/8.7/set-up-a-data-stream.html).

I agree on hard-coded though, I will fix that.

Reply:
> Or, do you mean check that datasets is a valid elastic search index?

This is what I meant. We'll have to test this out - prod ES is version 5.4.0.

Reply:
https://www.elastic.co/guide/en/elasticsearch/reference/5.4/docs-index_.html#index-creation
I haven't tested yet but it looks like it was true back then, too.

Reply:
Worked for me - didn't need to create `datasets` index manually before adding datasets.